### PR TITLE
Utkan 3 underlay interface

### LINKED ISSUE #3: Underlay Interface
* To Implement the Underlay Interface According to [this](https://github.com/yhassanzadeh13/SkipGraphNode/issues/1#issuecomment-632969218).
* The Function `SendRequest` takes:
   * `Recipient`: string, address of recipient node.
   * `RequestType`: which denotes which service or function to be invoked, can be for example an enum type.
   * RequestParameters, for now let's make this an interface, for which we can later have a class implementing that interface, and defining its own set of parameters according to the type of request, and depending on which serialization method we use, we can then adapts these classes to be serialized later.
   * `RequestResponse`: can be treated the same way as `RequestParameters`
* `ConnectionAdapter` can be an interface for which any RPC service we will use will have its adapter implementing this interface.

### LINKED ISSUE #11: Testing of Underlay RMI Implementation
* To add unit tests for the RMI Implementation of the underlay

### PR DESCRIPTION
**Pull request by utkn**
Related issue: #3 
* I have created the appropriate interfaces and classes.
* I have also implemented an example Java RMI Underlay using our new interface. 
* The example implementation is to demonstrate my understanding of the implementation of the underlay interface. Please let me know if there are any problems with my understanding.

In general, please let me know if any changes & additions are necessary.

### CODE REVIEW

**Comment by NazirNayal8** [comment]
Greetings @utkn , Thank you for the implementation. This version seems to work very well, but I was thinking of a small modification and I would like your opinion. Instead of having `Underlay` being inherited by `JavaRMIUnderlay`, we can have the `Underlay` as a standalone class, and only keep the `JavaRMIAdapter` which can be plugged in place of `connectionAdapter` in `Underlay`, and we can have the implementation of `sendMessage` function directly implemented in `Underlay` and it uses the functions provided by `ConnectionAdapter` Interface. Please tell me if my idea is not clear so that I can elaborate on it further

**Comment by utkn** [comment]
Dear @NazirNayal8,

I think the changes that you have proposed make a lot of sense. I have made the following changes as your suggestions:
* Converted Underlay into a concrete class. 
* ConnectionAdapter handles initialization & termination & connecting to a remote adapter.
* For different RMI implementations, we only need to implement the ConnectionAdapter interface.

By Sunday, I will add more comments (i.e. JavaDoc) to my code. Again, please let me know if you have any more suggestions, comments, or if there was any misunderstanding on my part.

Thanks,

Utkan

**Comment by NazirNayal8** [comment]
Greetings @utkn , Thank you for your modifications, this is exactly what I meant. In this case I will merge these modifications right after you add the comments.

File: src/underlay/RequestType.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package underlay;
+
+// Possible request types for a
+public enum RequestType {
+    SearchByNameID,
+    SearchByNumID,
+    NameIDLevelSearch,
+    UpdateLeftNode,
+    UpdateRightNode
+}
```
Comment (thep2p) [inline src/underlay/RequestType.java]: @utkn very nice idea 👏

**Comment by utkn** [comment]
Dear @NazirNayal8,

I have added the comments, and I believe the branch is ready to merge.

Thanks,

Utkan

**Comment by utkn** [comment]
Unit tests added as addressed in #11.
* Changing the connection adapter of the underlay dynamically is not possible now.
* An `Underlay` uses the `ConnectionAdapter` instance returned from the `Underlay.defaultAdapter()` method.
* Created a `test` module.
* Included `Junit5` as a dependency.
* Unit tests for `JavaRMIAdapter` added.
* Unit tests for `Underlay` added.
* All the necessary comments are added.

**Comment by NazirNayal8** [comment]
@utkn , Thank you very much all looks good

**Review by NazirNayal8 — APPROVED** [review]

**Pull request merged**